gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedgoogleHadoopFSInputStream.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2013 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.fs.gcs;
+
+import com.google.cloud.hadoop.gcsio.FileInfo;
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageReadOptions;
+import java.io.IOException;
+import java.net.URI;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.statistics.IOStatistics;
+import org.apache.hadoop.fs.statistics.IOStatisticsSource;
+
+/** A class to update the input stream statistics of {@link GoogleHadoopFSInputStreamBase} */
+public class InstrumentatedgoogleHadoopFSInputStream extends GoogleHadoopFSInputStreamBase
+    implements IOStatisticsSource {
+
+  // Statistic tracker of the Input stream
+  private final GhfsInputStreamStatistics streamStatistics;
+
+  // IO Statistics of the current input stream
+  private final IOStatistics ioStatistics;
+
+  InstrumentatedgoogleHadoopFSInputStream(
+      GoogleHadoopFileSystemBase ghfs,
+      URI gcsPath,
+      GoogleCloudStorageReadOptions readOptions,
+      FileSystem.Statistics statistics)
+      throws IOException {
+    super(ghfs, gcsPath, readOptions, statistics);
+    this.streamStatistics =
+        ((InstrumentatedGoogleHadoopFileSystem) ghfs)
+            .getInstrumentation()
+            .newInputStreamStatistics(statistics);
+    this.ioStatistics = streamStatistics.getIOStatistics();
+  }
+
+  InstrumentatedgoogleHadoopFSInputStream(
+      GoogleHadoopFileSystemBase ghfs, FileInfo fileInfo, FileSystem.Statistics statistics)
+      throws IOException {
+    super(ghfs, fileInfo, statistics);
+    this.streamStatistics =
+        ((InstrumentatedGoogleHadoopFileSystem) ghfs)
+            .getInstrumentation()
+            .newInputStreamStatistics(statistics);
+    this.ioStatistics = streamStatistics.getIOStatistics();
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    streamStatistics.readOperationStarted(getPos(), 1);
+    int response = 0;
+    try {
+      response = super.read();
+    } catch (IOException e) {
+      streamStatistics.readException();
+      streamStatistics.readOperationCompleted(1, response);
```

Comment:
we are capturing the readOperationCompleted twice here, and consuming the IOException without propagating.

Can we throw the exception upstream after recording the exception metric ?

Reply:
Updated